File: python/caskada/_state.py
# This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
# Copyright (c) 2025, Victor Duarte
from __future__ import annotations

from collections.abc import Mapping
from typing import Any, cast

from ._contracts import OptionValidationError

type _StateCarrier = dict[str, Any]


def _capture_initial_state(initial_state: object) -> _StateCarrier:
    if not isinstance(initial_state, Mapping):
        raise OptionValidationError("initial_state must be a Mapping")
    try:
        state = dict(initial_state)
    except BaseException as cause:
        raise OptionValidationError(
            "initial_state could not be shallow-copied"
        ) from cause
    if any(type(key) is not str for key in state):
        raise OptionValidationError("initial_state keys must be exact strings")
    return cast(_StateCarrier, state)

File: typescript/state.ts
// This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
// Copyright (c) 2026, Victor Duarte

export function captureInitialState<State extends object>(initialState: Readonly<State>): State {
  if (typeof initialState !== 'object' || initialState === null || Array.isArray(initialState)) {
    throw new TypeError('initialState must be an object')
  }
  const prototype = Object.getPrototypeOf(initialState)
  if (prototype !== Object.prototype && prototype !== null) {
    throw new TypeError('initialState must be a plain object')
  }
  if (Reflect.ownKeys(initialState).some((key) => typeof key !== 'string')) {
    throw new TypeError('initialState keys must be strings')
  }
  return { ...initialState }
}

export function resolveState<State extends object>(resolve: (state: State) => void, state: State): void {
  if (typeof Reflect.get(state, 'then') !== 'function') {
    resolve(state)
    return
  }
  const previous = Object.getOwnPropertyDescriptor(state, 'then')
  try {
    Object.defineProperty(
      state,
      'then',
      previous !== undefined && 'value' in previous ? { ...previous, value: undefined } : { value: undefined, configurable: true },
    )
  } catch {
    throw new TypeError('final state with a callable then property must remain mutable')
  }
  try {
    resolve(state)
  } finally {
    if (previous === undefined) Reflect.deleteProperty(state, 'then')
    else Object.defineProperty(state, 'then', previous)
  }
}
